Regression with queries that use value variables causing panics

If you suspect this could be a bug, follow the template.

- What version of Dgraph are you using?

All experiments run with Dgraph Docker images from Docker Hub

dgraph/dgraph:master
dgraph/dgraph:v1.0.15-rc9
dgraph/dgraph:v1.0.14
dgraph/dgraph:v1.0.13
dgraph/dgraph:v1.0.12
dgraph/dgraph:v1.0.11

- Have you tried reproducing the issue with latest release?

Yes, see versions listed above.

- What is the hardware spec (RAM, OS)?

1 Zero, 1 Alpha - 4 GB - Linux

3 Zero, 3 Alpha - Lots of memory - Linux

- Steps to reproduce the issue (command/config used to run Dgraph).

Commands to start local 1 Zero, 1 Alpha cluster.
```
$ docker run --rm -it --net=host dgraph/dgraph:${VERSION} dgraph zero
$ docker run --rm -it --net=host dgraph/dgraph:${VERSION} dgraph alpha --lru_mb=4096
```

Schema.
```
<number>: int @index(int) .
```

Query to get the highest available number value.
```
{
  var(func: has(number)) { 
    number as number 
  }
  var() { 
    highest as max(val(number)) 
  }

  all(func: eq(number, val(highest))) { 
    uid 
    number 
  }
}
```

Another query to get the highest available number value.
```
{
  var(func: has(number, orderdesc: number, first: 1) 
  {
    highest_number as number
  }

  all(func: eq(number, val(highest_number))) 
  { 
    uid 
  }
}
```

- Expected behaviour and actual result.

The queries above, when run against a Dgraph cluster without any loaded data, should yield an empty list.

dgraph/dgraph:v1.0.11 returns an empty list appropriately

all other versions panic and the alpha crashes.

Error message yielded in client.
```
E           grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
E               status = StatusCode.UNAVAILABLE
E               details = "Socket closed"
E               debug_error_string = "{"created":"@1558883325.760513047","description":"Error received from peer ipv4:127.0.0.1:9080","file":"src/core/lib/surface/call.cc","file_line":1041,"grpc_message":"Socket closed","grpc_status":14}"
E           >
```

Error message from alpha logs.
```
panic: assignment to entry in nil map

goroutine 291 [running]:
github.com/dgraph-io/dgraph/query.(*SubGraph).fillVars(0xc00039d200, 0xc000634480, 0x1, 0x1)
        /ext-go/1/src/github.com/dgraph-io/dgraph/query/query.go:1721 +0x5a7
github.com/dgraph-io/dgraph/query.(*SubGraph).recursiveFillVars(0xc00039d200, 0xc000634480, 0xc00060b020, 0x1)
        /ext-go/1/src/github.com/dgraph-io/dgraph/query/query.go:1660 +0x39
github.com/dgraph-io/dgraph/query.(*QueryRequest).ProcessQuery(0xc0100218f8, 0x174bac0, 0xc0002b9a40, 0x1, 0x1)
        /ext-go/1/src/github.com/dgraph-io/dgraph/query/query.go:2676 +0x68a
github.com/dgraph-io/dgraph/query.(*QueryRequest).Process(0xc0100218f8, 0x174bac0, 0xc0002b9a40, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        /ext-go/1/src/github.com/dgraph-io/dgraph/query/query.go:2770 +0x72
github.com/dgraph-io/dgraph/edgraph.(*Server).doQuery(0x2217178, 0x174bac0, 0xc0002bd830, 0xc00020e870, 0xc000309980, 0x0, 0x0)
        /ext-go/1/src/github.com/dgraph-io/dgraph/edgraph/server.go:652 +0x772
github.com/dgraph-io/dgraph/edgraph.(*Server).Query(0x2217178, 0x174bac0, 0xc0002b9830, 0xc00020e870, 0x2217178, 0xc0002b9320, 0x14aa080)
        /ext-go/1/src/github.com/dgraph-io/dgraph/edgraph/server.go:561 +0x98
github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/dgo/protos/api._Dgraph_Query_Handler(0x155bf60, 0x2217178, 0x174bac0, 0xc0002b9830, 0xc00020e820, 0x0, 0x0, 0x0, 0xc00023c0a0, 0x91)
        /ext-go/1/src/github.com/dgraph-io/dgraph/vendor/github.com/dgraph-io/dgo/protos/api/api.pb.go:1917 +0x23e
github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc.(*Server).processUnaryRPC(0xc0005d5c80, 0x1751f40, 0xc00022a900, 0xc00021e800, 0xc00031b980, 0x2153c78, 0x0, 0x0, 0x0)
        /ext-go/1/src/github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc/server.go:972 +0x4a2
github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc.(*Server).handleStream(0xc0005d5c80, 0x1751f40, 0xc00022a900, 0xc00021e800, 0x0)
        /ext-go/1/src/github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc/server.go:1252 +0xe02
github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc00042c800, 0xc0005d5c80, 0x1751f40, 0xc00022a900, 0xc00021e800)
        /ext-go/1/src/github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc/server.go:691 +0x9f
created by github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc.(*Server).serveStreams.func1
        /ext-go/1/src/github.com/dgraph-io/dgraph/vendor/google.golang.org/grpc/server.go:689 +0xa1
```

Issues that seem relevant:

- https://github.com/dgraph-io/dgraph/issues/713
- https://github.com/dgraph-io/dgraph/issues/3041
- https://github.com/dgraph-io/dgraph/issues/3201

--- a/query/common_test.go
+++ b/query/common_test.go
@@ -220,13 +220,13 @@ type SchoolInfo {
 	abbr: string
 	school: [uid]
 	district: [uid]
-	state: [uid]	
+	state: [uid]
 	county: [uid]
 }
 
 type User {
 	name: string
-	password: password	
+	password: password
 }
 
 type Node {
@@ -271,6 +271,7 @@ year                           : int .
 previous_model                 : uid @reverse .
 created_at                     : datetime @index(hour) .
 updated_at                     : datetime @index(year) .
+number                         : int @index(int) .
 `
 
 func populateCluster() {

--- a/query/query.go
+++ b/query/query.go
@@ -343,6 +343,11 @@ func aggWithVarFieldName(pc *SubGraph) string {
 	return fieldName
 }
 
+func emptyIneqFnWithVar(sg *SubGraph) bool {
+	return sg.SrcFunc != nil && isInequalityFn(sg.SrcFunc.Name) && len(sg.SrcFunc.Args) == 0 &&
+		len(sg.Params.NeedsVar) > 0
+}
+
 func addInternalNode(pc *SubGraph, uid uint64, dst outputNode) error {
 	sv, ok := pc.Params.uidToVal[uid]
 	if !ok || sv.Value == nil {
@@ -2671,8 +2676,7 @@ func (req *Request) ProcessQuery(ctx context.Context) (err error) {
 			// 1. It just does aggregation and math functions which is when sg.Params.IsEmpty is true.
 			// 2. Its has an inequality fn at root without any args which can happen when it uses
 			// value variables for args which don't expand to any value.
-			if sg.Params.IsEmpty ||
-				sg.SrcFunc != nil && isInequalityFn(sg.SrcFunc.Name) && len(sg.SrcFunc.Args) == 0 {
+			if sg.Params.IsEmpty || emptyIneqFnWithVar(sg) {
 				errChan <- nil
 				continue
 			}

--- a/query/query.go
+++ b/query/query.go
@@ -344,9 +344,6 @@ func aggWithVarFieldName(pc *SubGraph) string {
 }
 
 func addInternalNode(pc *SubGraph, uid uint64, dst outputNode) error {
-	if len(pc.Params.uidToVal) == 0 {
-		return nil
-	}
 	sv, ok := pc.Params.uidToVal[uid]
 	if !ok || sv.Value == nil {
 		return nil
@@ -1709,21 +1706,6 @@ func (sg *SubGraph) fillVars(mp map[string]varValue) error {
 				return errors.Errorf("Wrong variable type encountered for var(%v) %v.", v.Name, v.Typ)
 
 			default:
-				// This var does not match any uids or vals but we are still trying to access it.
-				if v.Typ == gql.ValueVar {
-					// * * * * * * * * * * * * * * * * * * *
-					// Default value vars
-					// * * * * * * * * * * * * * * * * * * *
-					//
-					// Provide a default value for valueVarAggregation() to eval val().
-					// This is a noop for aggregation funcs that would fail.
-					// The zero aggs won't show because there are no uids matched.
-					//
-					// NOTE: If you need to make type assertions that might involve
-					// default value vars, use `Safe()` func and not val.Value directly.
-					mp[v.Name].Vals[0] = types.Val{}
-					sg.Params.uidToVal = mp[v.Name].Vals
-				}
 			}
 		}
 	}
@@ -1749,7 +1731,9 @@ func (sg *SubGraph) replaceVarInFunc() error {
 			continue
 		}
 		if len(sg.Params.uidToVal) == 0 {
-			return errors.Errorf("No value found for value variable %q", arg.Value)
+			// This means that the variable didn't have any values and hence there is nothing to add
+			// to args. Setting isEmpty to true would mean that they query is not processed.
+			break
 		}
 		// We don't care about uids, just take all the values and put as args.
 		// There would be only one value var per subgraph as per current assumptions.
@@ -2683,8 +2667,12 @@ func (req *Request) ProcessQuery(ctx context.Context) (err error) {
 			hasExecuted[idx] = true
 			numQueriesDone++
 			idxList = append(idxList, idx)
-			// Doesn't need to be executed as it just does aggregation and math functions.
-			if sg.Params.IsEmpty {
+			// A query doesn't need to be executed if
+			// 1. It just does aggregation and math functions which is when sg.Params.IsEmpty is true.
+			// 2. Its has an inequality fn at root without any args which can happen when it uses
+			// value variables for args which don't expand to any value.
+			if sg.Params.IsEmpty ||
+				sg.SrcFunc != nil && isInequalityFn(sg.SrcFunc.Name) && len(sg.SrcFunc.Args) == 0 {
 				errChan <- nil
 				continue
 			}

--- a/query/query.go
+++ b/query/query.go
@@ -1661,6 +1661,9 @@ func (sg *SubGraph) populateFacetVars(doneVars map[string]varValue, sgPath []*Su
 	return nil
 }
 
+// recursiveFillVars fills the value of variables before a query is to be processed using the result
+// of the values (doneVars) which were computed by other queries which could be run before this
+// query.
 func (sg *SubGraph) recursiveFillVars(doneVars map[string]varValue) error {
 	err := sg.fillVars(doneVars)
 	if err != nil {
@@ -1693,7 +1696,7 @@ func (sg *SubGraph) fillVars(mp map[string]varValue) error {
 			case (v.Typ == gql.AnyVar || v.Typ == gql.UidVar) && l.Uids != nil:
 				lists = append(lists, l.Uids)
 
-			case (v.Typ == gql.AnyVar || v.Typ == gql.ValueVar) && len(l.Vals) != 0:
+			case (v.Typ == gql.AnyVar || v.Typ == gql.ValueVar):
 				// This should happen only once.
 				// TODO: This allows only one value var per subgraph, change it later
 				sg.Params.uidToVal = l.Vals
@@ -2398,7 +2401,10 @@ func (sg *SubGraph) sortAndPaginateUsingFacet(ctx context.Context) error {
 }
 
 func (sg *SubGraph) sortAndPaginateUsingVar(ctx context.Context) error {
-	if len(sg.Params.uidToVal) == 0 {
+	// nil has a different from an initialized map of zero length here. If the variable didn't return
+	// any values then uidToVal would be an empty with zero length. If the variable was used before
+	// definition, the uidToVal would be nil.
+	if sg.Params.uidToVal == nil {
 		return errors.Errorf("Variable: [%s] used before definition.", sg.Params.Order[0].Attr)
 	}
 

--- a/query/query.go
+++ b/query/query.go
@@ -2401,9 +2401,9 @@ func (sg *SubGraph) sortAndPaginateUsingFacet(ctx context.Context) error {
 }
 
 func (sg *SubGraph) sortAndPaginateUsingVar(ctx context.Context) error {
-	// nil has a different from an initialized map of zero length here. If the variable didn't return
-	// any values then uidToVal would be an empty with zero length. If the variable was used before
-	// definition, the uidToVal would be nil.
+	// nil has a different meaning from an initialized map of zero length here. If the variable
+	// didn't return any values then uidToVal would be an empty with zero length. If the variable
+	// was used before definition, uidToVal would be nil.
 	if sg.Params.uidToVal == nil {
 		return errors.Errorf("Variable: [%s] used before definition.", sg.Params.Order[0].Attr)
 	}

--- a/query/query.go
+++ b/query/query.go
@@ -343,7 +343,7 @@ func aggWithVarFieldName(pc *SubGraph) string {
 	return fieldName
 }
 
-func emptyIneqFnWithVar(sg *SubGraph) bool {
+func isEmptyIneqFnWithVar(sg *SubGraph) bool {
 	return sg.SrcFunc != nil && isInequalityFn(sg.SrcFunc.Name) && len(sg.SrcFunc.Args) == 0 &&
 		len(sg.Params.NeedsVar) > 0
 }
@@ -1662,8 +1662,7 @@ func (sg *SubGraph) populateFacetVars(doneVars map[string]varValue, sgPath []*Su
 }
 
 // recursiveFillVars fills the value of variables before a query is to be processed using the result
-// of the values (doneVars) which were computed by other queries which could be run before this
-// query.
+// of the values (doneVars) computed by other queries that were successfully run before this query.
 func (sg *SubGraph) recursiveFillVars(doneVars map[string]varValue) error {
 	err := sg.fillVars(doneVars)
 	if err != nil {
@@ -1714,6 +1713,7 @@ func (sg *SubGraph) fillVars(mp map[string]varValue) error {
 				return errors.Errorf("Wrong variable type encountered for var(%v) %v.", v.Name, v.Typ)
 
 			default:
+				glog.V(3).Infof("Warning: reached default case in fillVars for var: %v", v.Name)
 			}
 		}
 	}
@@ -2682,7 +2682,7 @@ func (req *Request) ProcessQuery(ctx context.Context) (err error) {
 			// 1. It just does aggregation and math functions which is when sg.Params.IsEmpty is true.
 			// 2. Its has an inequality fn at root without any args which can happen when it uses
 			// value variables for args which don't expand to any value.
-			if sg.Params.IsEmpty || emptyIneqFnWithVar(sg) {
+			if sg.Params.IsEmpty || isEmptyIneqFnWithVar(sg) {
 				errChan <- nil
 				continue
 			}

--- a/query/query.go
+++ b/query/query.go
@@ -1740,7 +1740,7 @@ func (sg *SubGraph) replaceVarInFunc() error {
 		}
 		if len(sg.Params.uidToVal) == 0 {
 			// This means that the variable didn't have any values and hence there is nothing to add
-			// to args. Setting isEmpty to true would mean that they query is not processed.
+			// to args.
 			break
 		}
 		// We don't care about uids, just take all the values and put as args.

--- a/query/query0_test.go
+++ b/query/query0_test.go
@@ -1809,7 +1809,7 @@ func TestNonFlattenedResponse(t *testing.T) {
 		me(func: eq(name@en, "Baz Luhrmann")) {
 			uid
 			director.film {
-				name@en	
+				name@en
 			}
 		}
 	}`

--- a/query/query1_test.go
+++ b/query/query1_test.go
@@ -1241,6 +1241,60 @@ func TestAggregateRootError(t *testing.T) {
 	require.Contains(t, err.Error(), "Only aggregated variables allowed within empty block.")
 }
 
+func TestAggregateEmpty1(t *testing.T) {
+	query := `
+	{
+		var(func: has(number)) {
+			number as number
+		}
+		var() {
+			highest as max(val(number))
+		}
+
+		all(func: eq(number, val(highest))) {
+			uid
+			number
+		}
+	}`
+
+	js := processQueryNoErr(t, query)
+	require.JSONEq(t, `{"data": {"all":[]}}`, js)
+}
+
+func TestAggregateEmpty2(t *testing.T) {
+	query := `
+		{
+			var(func: has(number))
+			{
+				highest_number as number
+			}
+			all(func: eq(number, val(highest_number)))
+			{
+				uid
+			}
+		}
+	`
+	js := processQueryNoErr(t, query)
+	require.JSONEq(t, `{"data": {"all":[]}}`, js)
+}
+
+func TestAggregateEmpty3(t *testing.T) {
+	query := `
+		{
+			var(func: has(number))
+			{
+				highest_number as number
+			}
+			all(func: ge(number, val(highest_number)))
+			{
+				uid
+			}
+		}
+	`
+	js := processQueryNoErr(t, query)
+	require.JSONEq(t, `{"data": {"all":[]}}`, js)
+}
+
 func TestFilterLang(t *testing.T) {
 	// This tests the fix for #1334. While getting uids for filter, we fetch data keys when number
 	// of uids is less than number of tokens. Lang tag was not passed correctly while fetching these

